File: Assignment 4/C/08. pair wise swap elements in a linked list.c
// input : 1 2 3 4
// output: 2 1 4 3
// trick - just swap tha values :)
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    int val;
    struct node* next;
}node;

node* createNode(int n){
    node* t = (node*)malloc(sizeof(node));
    t->val = n;
    t->next = NULL;
    return t;
}

void push(node* head,int n){
    node* t = (node*)malloc(sizeof(node));
    if(t==NULL){
        printf("no space for %d",n);
        return;
    }
    t->val = n;
    t->next = NULL;
    while(head->next){
        head = head->next;
    }
    head->next = t;
}

node* swap(node* head){
    node* temp = head;
    while(temp && temp->next){
        int k = temp->val;
        temp->val = temp->next->val;
        temp->next->val = k;

        temp = temp->next->next;
    }
    return head;
}

void traversal(node* head){
    node* ptr = head;
    while(ptr){
        printf("%d --> ",ptr->val);
        ptr = ptr->next;
    }
    printf("NULL\n");
}

int main(){
    int n;
    printf("no of elements: ");
    scanf("%d",&n);
    int arr[n];
    printf("enter %d value: ",n);
    for(int i=0; i<n; i++){
        scanf("%d",&arr[i]);
    }

    node* head = createNode(arr[0]);
    for(int i=1; i<n; i++){
        push(head,arr[i]);
    }
    traversal(head);

    swap(head);
    traversal(head);

    return 0;
}


//OUTPUT:

// no of elements: 4
// enter 4 value: 1 2 3 4
// 1 --> 2 --> 3 --> 4 --> NULL
// 2 --> 1 --> 4 --> 3 --> NULL

File: Assignment 4/Python/08. pair wise swap elements in a linked list .py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def create_node(n):
    return Node(n)

def push(head, n):
    new_node = Node(n)
    if not head:
        return new_node
    ptr = head
    while ptr.next:
        ptr = ptr.next
    ptr.next = new_node
    return head

def swap(head):
    temp = head
    while temp and temp.next:
        temp.val, temp.next.val = temp.next.val, temp.val
        temp = temp.next.next
    return head

def traversal(head):
    ptr = head
    while ptr:
        print(f"{ptr.val} --> ", end="")
        ptr = ptr.next
    print("NULL")

if __name__ == "__main__":
    n = int(input("Number of elements: "))
    arr = [int(x) for x in input(f"Enter {n} values: ").split()]

    head = create_node(arr[0])
    for i in range(1, n):
        head = push(head, arr[i])

    traversal(head)
    head = swap(head)
    traversal(head)


# OUTPUT:
# Number of elements: 4
# Enter 4 values: 1 2 3 4
# 1 --> 2 --> 3 --> 4 --> NULL
# 2 --> 1 --> 4 --> 3 --> NULL
